### ISSUE TITLE
Coarse `Lq` tiling produces wrong results for GQA flash attention with folded `G * Lq` dimension

### ISSUE THREAD

**Issue by cyang49** [torch-spyre/torch-spyre#3098]
### What happened?

**Description**

`flash_spyre` in `test-spyre-scripts/test_flash_gqa.py` produces correct results without an `Lq` tiling hint, but mismatches when enabling:

```python
with spyre_hint(tiles={"Lq": Lq // q_block_size}):
```

The implementation folds queries from `[B, Hq, Lq, D]` into `[B, Hkv, G * Lq, D]` for GQA, with `Hq = Hkv * G`.

Without the `Lq` hint, the test passes. With the hint enabled around the attention body, the result mismatches:

```text
Mismatched elements: ~1,017,611 / 16,777,216
Greatest absolute difference: ~3.49
```

`flash_cpu` was also checked against:

```python
torch.nn.functional.scaled_dot_product_attention(..., enable_gqa=True)
```

and matched within tolerance, so this does not appear to be a reference issue.

**Artifact Clue**

In the hinted compiler artifact, the `Lq` tile creates:

```python
LoopSpec(count=sympify("4"))
```

and many ops inside the loop shrink the folded dimension from `16384` to `4096`. The query load looks suspicious:

```python
device_size=[32, 4, 4096, 2, 64]
device_coordinates=[c0, 0, c1, floor(c2/64), Mod(c2, 64)]
```

Here `c0` appears to be the `Hkv` loop but is used directly as the original `Hq` coordinate. For GQA, the original query head should account for both `Hkv` and `G`, roughly `Hq = Hkv * G + g`.

A mismatch diagnostic also supports this: heads `0..3` are correct, while heads `4..31` mismatch, suggesting the folded `[Hkv, G, Lq] -> [Hq, Lq]` mapping is lost or mis-indexed under coarse tiling.

**Suspected Cause**

`coarse_tile`/hint propagation appears to tile `Lq` on the folded `G * Lq` loop, but downstream indexing for the original query view does not preserve the compound named-dim decomposition `["Hkv", "G", "Lq"]`. This causes the tiled kernel to read query heads as if `Hq == Hkv`, dropping the `G` contribution.

### What did you expect to happen?

Tiling or not, it should not create numerical error larger than the tolerance.

### How can we reproduce it?

Tile on a folded dimension.

### Any environmental details we need to know?

No

### Relevant log output

```Shell

```

**Comment by cyang49** [comment]
**Debugging Report**

Issue: `flash_spyre` GQA result mismatch when enabling coarse tiling on `Lq`.

Focused repro:

```text
B=1, Hq=8, Hkv=2, G=4, D=128
Lq=1024, Lk=256
tiles={"Lq": 4}
```

Reference used: CPU `torch.nn.functional.scaled_dot_product_attention(..., enable_gqa=True)` and a reduced scores-only `torch.matmul` repro.

**Observed Behavior**

Scores-only repro shows the bug before softmax/output accumulation:

```text
split=2: bad 0 / 2097152, max diff ~0.0195
split=4: bad 1018981 / 2097152, max diff 8.375
```

For split-4, all bad scores are in `Hkv=1`:

```text
per_hkv [0, 1018981]
```

Full flash repro similarly fails only heads 4-7, i.e. the second `Hkv` group.

Adding `.contiguous()` or `.clone()` after:

```python
queries_folded = queries.view(B, Hkv, G * Lq, D)
```

did not change results; those ops were optimized away.

**Key Artifact Finding**

For failing split-4, the first Q scaling `mul` starts with a correct pre-alignment TensorArg:

```text
size   [8, 1024, 2, 1, 64]
coords [4*c0, c1, floor(c2/64), 0, Mod(c2,64)]
```

Here `c0` is `Hkv`. The expression `4*c0` correctly maps `Hkv=0/1` to physical Q heads `0/4`. The unrolled Lq tile offset supplies the G lane.

After `align_tensors()`, this becomes:

```text
size   [8, 4, 1024, 2, 64]
coords [c0, 0, c1, floor(c2/64), Mod(c2,64)]
```

The `size=4, coord=0` dimension represents the stride/gap from the original `4*c0`.

**Why BackGap Is Wrong**

SuperDSC derives logical dim order from coordinate symbols in `_get_device_dim_order()`:

```text
coords [c0, 0, c1, floor(c2/64), Mod(c2,64)]
```

Since the `size=4` dimension has coordinate `0`, it is dropped. SuperDSC sees:

```text
dim_order = [out, x, mb]
```

Then `_create_sdsc_tensors()` uses positional indexing:

```python
stride_idx = stride_dim_order.index(dim)
dev_dim_size = arg.device_size[-stride_idx - 2]
```

Because the constant physical dimension was dropped from `dim_order` but remains in `device_size`, `mb` gets paired with the wrong physical slot:

```text
mb: dev_dim_size=4, it_dim_size=2
```

So the generated SDSC contains:

```text
backGap {'mb': {'-1': '2'}}
```

But the expected gap is on the folded query/Lq axis:

```text
backGap {'x': {'-1': '3072'}}
```

where `3072 = G * Lq - Lq = 4096 - 1024`.

**Proof Experiment**

I temporarily patched `simplify_op_spec()` to coalesce the anonymous constant G dimension back into the folded Lq dimension after `align_tensors()`.

With that proof patch:

```text
scores-only split=4:
bad 0 / 2097152
max diff ~0.0195
per_hkv [0, 0]
```

The SDSC changed from:

```text
backGap {'mb': {'-1': '2'}}
```

to:

```text
backGap {'x': {'-1': '3072'}}
```

Reduced full flash also passed against CPU SDPA:

```text
bad 0 / 1048576
max diff ~0.0068
per_head [0,0,0,0,0,0,0,0]
```

The proof patch was reverted; tracked files are clean.

**Conclusion**

This is not an attention math bug. It is a layout/codegen bug exposed by GQA folding plus coarse tiling.

Root cause:

```text
align_tensors() preserves a view-induced stride multiplier as an anonymous
constant physical dimension, but SuperDSC backGap assignment assumes physical
device dimensions can be matched positionally to symbolic logical dims.
```

Coarse tiling exposes the issue because within each Lq tile the G lane is no longer represented by a loop symbol; it is supplied by the unrolled tile base offset. That leaves the `G=4` factor as `coord=0, size=4`, which SuperDSC mishandles.

Likely fix direction: make SuperDSC’s backGap/stride mapping robust to anonymous constant physical dimensions, or avoid emitting this decomposed constant dimension for folded GQA-style views when it should remain coalesced with the folded query axis.

**Comment by cyang49** [comment]
Minimal reproducer

```
import torch
from torch_spyre._inductor import spyre_hint
from torch_spyre._inductor.propagate_named_dims import (
    declare_tensor_dim,
    name_tensor_dims,
)

SEED = 42

B, Hq, Hkv, Lq, D = 1, 8, 2, 1024, 128
G = Hq // Hkv
LQ_TILES = 4


def folded_gqa_pointwise(x):
    x = x.view(B, Hkv, G * Lq, D)
    with spyre_hint(tiles={"Lq": LQ_TILES}):
        return (x + 1).contiguous()


def declare_dims():
    for name, size in {
        "Hkv": Hkv,
        "G": G,
        "Lq": Lq,
        "D": D,
    }.items():
        declare_tensor_dim(name, size)


if __name__ == "__main__":
    torch.manual_seed(SEED)

    x = torch.randn(B, Hq, Lq, D, dtype=torch.float16)
    expected = folded_gqa_pointwise(x)

    x_spyre = x.to(device="spyre")
    declare_dims()
    name_tensor_dims(x_spyre, ["Hkv", "G", "Lq", "D"])

    actual = torch.compile(folded_gqa_pointwise)(x_spyre).cpu()
    torch.testing.assert_close(expected, actual, atol=0.1, rtol=0.1)
    print("SUCCESS")
```

**Comment by cyang49** [comment]
coarse tiling revamped 
need retesting moving target to Aug 2026
